packages/cloud_firestore/cloud_firestore_platform_interface/lib/cloud_firestore_platform_interface.dart

```diff
@@ -0,0 +1,219 @@
+/// changes. Extending this class (using `extends`) ensures that the subclass
+/// will get the default implementation, while platform implementations that
+/// `implements` this interface will be broken by newly added
+/// [CloudFirestorePlatform] methods.
+abstract class CloudFirestorePlatform {
+  /// Only mock implementations should set this to `true`.
+  ///
+  /// Mockito mocks implement this class with `implements` which is forbidden
+  /// (see class docs). This property provides a backdoor for mocks to skip the
+  /// verification that the class isn't implemented with `implements`.
+  @visibleForTesting
+  bool get isMock => false;
+
+  /// The default instance of [CloudFirestorePlatform] to use.
+  ///
+  /// Platform-specific plugins should override this with their own class
+  /// that extends [CloudFirestorePlatform] when they register themselves.
+  ///
+  /// Defaults to [MethodChannelCloudFirestore].
+  static CloudFirestorePlatform get instance => _instance;
+
+  static CloudFirestorePlatform _instance = MethodChannelCloudFirestore();
+
+  // TODO(amirh): Extract common platform interface logic.
+  // https://github.com/flutter/flutter/issues/43368
+  static set instance(CloudFirestorePlatform instance) {
+    if (!instance.isMock) {
+      try {
+        instance._verifyProvidesDefaultImplementations();
+      } on NoSuchMethodError catch (_) {
+        throw AssertionError(
+            'Platform interfaces must not be implemented with `implements`');
+      }
+    }
+    _instance = instance;
+  }
+
+  /// This method ensures that [CloudFirestorePlatform] isn't implemented with `implements`.
+  ///
+  /// See class docs for more details on why using `implements` to implement
+  /// [CloudFirestorePlatform] is forbidden.
+  ///
+  /// This private method is called by the [instance] setter, which should fail
+  /// if the provided instance is a class implemented with `implements`.
+  void _verifyProvidesDefaultImplementations() {}
+
+  // Actual API 
+  // Global
+  /// Removes any listener by its handle.
+  /// All handles must be unique across al types of listeners.
+  Future<void> removeListener(int handle) async {
+    throw UnimplementedError('removeListener() is not implemented');
+  }
+
+  // Firestore
+  Future<void> enablePersistence(String app, {@required bool enable}) async {
+    throw UnimplementedError('enablePersistence() is not implemented');
+  }
+
+  Future<void> settings(String app, {
+      bool persistenceEnabled,
+      String host,
+      bool sslEnabled,
+      int cacheSizeBytes,
+    }) async {
+    throw UnimplementedError('settings() is not implemented');
+  }
+
+  Future<Map<String, dynamic>> runTransaction(String app, {
+    @required PlatformTransactionHandler transactionHandler,
+    int transactionTimeout,
+  }) async {
+    throw UnimplementedError('runTransaction() is not implemented');
+  }
+
+  // Document Reference
+  Future<void> setDocumentReferenceData(String app, {
```

Comment:
It feels a little awkward we're merging the method namespace of every class in the Cloud Firestore API and using comments to disambiguate what classes the method names are referring to.

The method channel API uses `#` to separate classes and methods and that seemed a little cleaner.

We can't really split things out into `PlatformFirestore`, `PlatformDocumentReference` classes, etc. because we'll lose the protection against `implements` that only `FirestorePlatform` has.

I can live with this but since we're not allowed to make breaking changes we should make sure this is a pattern we want to propagate.

/cc @hterkelsen @amirh since this is pretty fundamental to how we're doing all the federated plugins.

Reply:
I tried to follow a naming convention to all these methods, so things like:

`DocumentReference#setData` -> `setDocumentReferenceData`

(I also did this to prevent naming collisions, with "snapshots", for example).

They all should be named alike. My second option was to separate each concern into their own **mixin** and apply those to the base class, but the naming alone should be enough. 

The comments you're seeing are remnants of the scaffolding that I once had when this class was empty :P